<!-- Google Fonts -->
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@700&display=swap" rel="stylesheet">

<div style="
    border-radius: 15px; 
    border: 2px solid #003366; 
    padding: 10px; 
    background: linear-gradient(135deg, #3a0ca3, #7209b7 30%, #f72585 80%);
    text-align: center; 
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.5);
">
    <h1 style="
        color: #fff;
        text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.7);
        font-weight: bold;
        margin-bottom: 10px;
        font-size: 36px;
        font-family: 'Roboto', sans-serif;
        letter-spacing: 1px;
    ">
        💳 Credit Risk Analysis 💸
    </h1>
</div>


# 📂 Overview

* **Background** <br>
  In the financial services industry, lending institutions face significant challenges in **managing credit risk** while maintaining profitability. Accurately assessing a borrower’s ability to repay is critical to **minimizing defaults** and ensuring sustainable lending practices.

* **Goal of the Project** <br>
  Using historical customer, financial, and credit behavior data, this project aims to build a predictive model that estimates the likelihood of a borrower defaulting on a loan. This enables financial institutions to **improve risk assessment**, **optimize loan approval decisions**, and **reduce potential losses**.

🎯 Key Features

**Target Variable**

* `loan_status`: *Loan status (e.g., fully paid, default, charged off)*

**Customer Demographics & Profile** <br>

* `person_age`: Age of the customer
* `gender`: Gender
* `marital_status`: Marital status
* `education_level`: Education level

**Financial & Employment Information** <br>

* `person_income`: Annual income
* `person_emp_length`: Employment length (in years)
* `employment_type`: Type of employment
* `person_home_ownership`: Home ownership status

**Loan Characteristics** <br>

* `loan_amnt`: Loan amount
* `loan_term_months`: Loan term (in months)
* `loan_int_rate`: Interest rate
* `loan_grade`: Loan risk grade (A–G)
* `loan_intent`: Loan purpose
* `loan_percent_income`: Loan-to-income percentage
* `loan_to_income_ratio`: Loan-to-income ratio

**Credit History** <br>

* `cb_person_cred_hist_length`: Credit history length
* `cb_person_default_on_file`: Historical default indicator

**Debt & Repayment Capacity** <br>

* `other_debt`: Other existing debts
* `debt_to_income_ratio`: Debt-to-income ratio

**Credit Behavior** <br>

* `open_accounts`: Number of open credit accounts
* `credit_utilization_ratio`: Credit utilization ratio
* `past_delinquencies`: Number of past delinquencies

**Geographic Information** <br>

* `country`: Country
* `state`: State/region
* `city`: City
* `city_latitude`: Latitude
* `city_longitude`: Longitude

📁 Files Provided

* `Credit_Risk_Dataset_Onyx_Data_September_25 - Credit Risk Data.csv`: Main dataset used for analysis and modeling

🚀 Project Objective

The primary objective of this project is to **perform exploratory data analysis (EDA)** and develop predictive models to classify whether a borrower is likely to default on a loan.

By leveraging these insights, financial institutions can:

* Improve **credit risk assessment**
* Enhance **loan approval strategies**
* Reduce **default rates and financial losses**

<!-- Include Google Fonts for a modern font -->
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@700&display=swap" rel="stylesheet">

# <span style="color:transparent;">Import Libraries</span>

<div style="
    border-radius: 15px; 
    border: 2px solid #003366; 
    padding: 10px; 
    background: linear-gradient(135deg, #3a0ca3, #7209b7 30%, #f72585 80%);
    text-align: center; 
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.5);
">
    <h1 style="
        color: #fff;
        text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.7);
        font-weight: bold;
        margin-bottom: 10px;
        font-size: 36px;
        font-family: 'Roboto', sans-serif;
        letter-spacing: 1px;
    ">
        Import Libraries
    </h1>
</div>


In [1]:
# ===============================
# Core Libraries
# ===============================
import pandas as pd              # Data handling
import numpy as np               # Numerical operations
import seaborn as sns            # Statistical plots
import matplotlib.pyplot as plt  # Visualization

# ===============================
# Project Utilities
# ===============================
from utils.logger import LoggerConfig
from utils.utils import Utils
from utils.config_loader import load_config
from config.config import CFG
from utils.colors_utils import ColorUtils
from utils.stats_utils import StatUtils
from utils.validation import DataValidator
from utils.feature_utils import FeatureUtils

# ===============================
# Modeling Utilities
# ===============================
from models.benchmark import *
from models.oof import *

# ===============================
# Hyperparameter Tuning
# ===============================
from models.catboost_tuning import *
from models.lgbm_tunning import *
from models.hgbc_tuning import *

# ===============================
# Explainability
# ===============================
from utils.explainability import *

# ===============================
# Notebook Display
# ===============================
from IPython.display import display, HTML

# ===============================
# Data Splitting & Preprocessing
# ===============================
from sklearn.model_selection import StratifiedShuffleSplit, StratifiedKFold

from sklearn.preprocessing import OneHotEncoder, RobustScaler, StandardScaler
from sklearn.impute import SimpleImputer

from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# ===============================
# Evaluation
# ===============================
from sklearn.metrics import classification_report

# ===============================
# Machine Learning Models
# ===============================
from sklearn.svm import LinearSVC, SVC
from sklearn.neighbors import KNeighborsClassifier

from sklearn.ensemble import (
    RandomForestClassifier,
    GradientBoostingClassifier,
    ExtraTreesClassifier,
    AdaBoostClassifier,
    BaggingClassifier,
    HistGradientBoostingClassifier
)

from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from catboost import CatBoostClassifier

from sklearn.neural_network import MLPClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression, RidgeClassifier

# ===============================
# Optimization
# ===============================
import optuna
optuna.logging.set_verbosity(optuna.logging.WARNING)

# ===============================
# Environment Settings
# ===============================
import warnings
warnings.filterwarnings("ignore")

pd.set_option("display.max_columns", 500)
pd.set_option("display.max_colwidth", None)

<!-- Include Google Fonts for a modern font -->
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@700&display=swap" rel="stylesheet">

# <span style="color:transparent;">Load Data</span>

<div style="
    border-radius: 15px; 
    border: 2px solid #003366; 
    padding: 10px; 
    background: linear-gradient(135deg, #3a0ca3, #7209b7 30%, #f72585 80%);
    text-align: center; 
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.5);
">
    <h1 style="
        color: #FFFFFF; 
        text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.7); 
        font-weight: bold; 
        margin-bottom: 5px; 
        font-size: 28px; 
        font-family: 'Roboto', sans-serif;
        letter-spacing: 1px;
    ">
        Load Data
    </h1>
</div>

In [2]:
# Initialize notebook logger
logger = LoggerConfig.setup_logger()

# Load datasets safely
df = Utils.read_csv_safe(path=CFG.DATA_PATH, logger=logger,
                         encodings=CFG.DEFAULT_ENCODING).pipe(Utils.clean_columns)

2026-04-23 22:11:41 | INFO     | Logging configured: level=DEBUG, format=colored
2026-04-23 22:11:41 | INFO     | Loaded file with encoding: utf-8 | Shape: (32581, 29)


<!-- Include Google Fonts for a modern font -->
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@700&display=swap" rel="stylesheet">

# <span style="color:transparent;">Data Preview and Info</span>

<div style="
    border-radius: 15px; 
    border: 2px solid #003366; 
    padding: 10px; 
    background: linear-gradient(135deg, #3a0ca3, #7209b7 30%, #f72585 80%);
    text-align: center; 
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.5);
">
    <h1 style="
        color: #FFFFFF; 
        text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.7); 
        font-weight: bold; 
        margin-bottom: 5px; 
        font-size: 28px; 
        font-family: 'Roboto', sans-serif;
        letter-spacing: 1px;
    ">
        Data Preview and Info
    </h1>
</div>


In [7]:
df["person_income"].min()

4000

In [8]:
df["person_income"].max()

6000000

In [15]:
df[df["person_emp_length"] > 40].head()

,client_ID,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length,gender,marital_status,education_level,country,state,city,city_latitude,city_longitude,employment_type,loan_term_months,loan_to_income_ratio,other_debt,debt_to_income_ratio,open_accounts,credit_utilization_ratio,past_delinquencies
0,CUST_00001,22,59000,RENT,123.0,PERSONAL,D,35000,16.02,1,0.59,Y,3,Male,Married,High School,Canada,Ontario,Toronto,43.6532,-79.3832,Self-employed,36,0.593220,8402.453850,0.735635,14,0.495557,0
210,CUST_00211,21,192000,MORTGAGE,123.0,VENTURE,A,20000,6.54,0,0.10,N,4,Male,Single,High School,UK,Wales,Swansea,51.6214,-3.9436,Full-time,36,0.104167,32653.092510,0.274235,6,0.482647,1
32355,CUST_32356,78,48000,RENT,41.0,MEDICAL,A,3000,7.51,0,0.06,N,25,Male,Married,High School,Canada,Ontario,Toronto,43.6532,-79.3832,Part-time,24,0.062500,4628.510849,0.158927,13,0.309243,1


In [5]:
# Data Preview
display(df.head())

,client_ID,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length,gender,marital_status,education_level,country,state,city,city_latitude,city_longitude,employment_type,loan_term_months,loan_to_income_ratio,other_debt,debt_to_income_ratio,open_accounts,credit_utilization_ratio,past_delinquencies
0,CUST_00001,22,59000,RENT,123.0,PERSONAL,D,35000,16.02,1,0.59,Y,3,Male,Married,High School,Canada,Ontario,Toronto,43.6532,-79.3832,Self-employed,36,0.593220,8402.453850,0.735635,14,0.495557,0
1,CUST_00002,21,9600,OWN,5.0,EDUCATION,B,1000,11.14,0,0.10,N,2,Female,Divorced,Master,Canada,Ontario,Toronto,43.6532,-79.3832,Full-time,36,0.104167,1607.802794,0.271646,10,0.585436,3
2,CUST_00003,25,9600,MORTGAGE,1.0,MEDICAL,C,5500,12.87,1,0.57,N,3,Female,Married,Master,UK,Wales,Swansea,51.6214,-3.9436,Full-time,36,0.572917,2760.505633,0.860469,14,0.750732,0
3,CUST_00004,23,65500,RENT,4.0,MEDICAL,C,35000,15.23,1,0.53,N,2,Male,Married,Bachelor,Canada,BC,Vancouver,49.2827,-123.1207,Part-time,12,0.534351,7155.286150,0.643592,15,0.379333,0
4,CUST_00005,24,54400,RENT,8.0,MEDICAL,C,35000,14.27,1,0.55,Y,4,Female,Single,Bachelor,USA,New York,Buffalo,42.8864,-78.8784,Part-time,36,0.643382,15626.153440,0.930628,4,0.228103,0


In [ ]:
# Display information about the DataFrames
df.info()

<div style="
    background: linear-gradient(135deg, #1f2a44, #2c3e50);
    color: #ecf0f1;
    padding: 25px;
    border-radius: 12px;
    margin: 25px 0;
    box-shadow: 0 6px 18px rgba(0,0,0,0.25);
">
<h2 style="color:#f9e79f; margin-top:0;">📊 Insights from Initial Data Exploration</h2>

<b>Dataset Size and Structure</b>  
<ul>
<li>The dataset contains <b>32,581 records</b> with <b>29 features</b>.</li>
<li>The data is moderately sized and suitable for both <b>EDA</b> and <b>machine learning modeling</b>.</li>
</ul>

<b>Feature Overview</b>
<ul>
<li>The dataset includes comprehensive information for credit risk assessment:</li>
<ul>
<li><b>Numerical features:</b> <code>person_age</code>, <code>person_income</code>, <code>loan_amnt</code>, <code>loan_int_rate</code>, <code>loan_percent_income</code>, <code>debt_to_income_ratio</code>, <code>credit_utilization_ratio</code>, <code>open_accounts</code>, <code>past_delinquencies</code>.</li>
<li><b>Categorical features:</b> <code>person_home_ownership</code>, <code>loan_intent</code>, <code>loan_grade</code>, <code>gender</code>, <code>marital_status</code>, <code>education_level</code>, <code>employment_type</code>, <code>cb_person_default_on_file</code>, <code>location-related features</code>.</li>
</ul>
<li>The target variable is <code>loan_status</code>, represented as a <b>binary outcome (0/1)</b>, indicating loan repayment status.</li>
</ul>

<b>Data Quality and Completeness</b>  
<ul>
<li>Most features are complete, but some missing values are observed:</li>
<ul>
<li><code>person_emp_length</code> has missing values (~2.7%).</li>
<li><code>loan_int_rate</code> has notable missing values (~9.6%).</li>
</ul>
<li>Data types are consistent:</li>
<ul>
<li><b>Numerical:</b> int64 and float64</li>
<li><b>Categorical:</b> object (string)</li>
</ul>
</ul>

</div>

In [ ]:
# Numerical Features
num_features = ["person_age", "person_income", "person_emp_length", "loan_amnt", "loan_int_rate", "loan_percent_income", 
                "loan_to_income_ratio", "cb_person_cred_hist_length", "other_debt",  "debt_to_income_ratio", "open_accounts", 
                "credit_utilization_ratio"]

# Categorical Features
cat_features = ["person_home_ownership", "loan_intent", "loan_grade", "cb_person_default_on_file", "gender", "marital_status", 
                "education_level", "employment_type", "country", "state", "city", "loan_term_months", "past_delinquencies"]

# Data Describe
cm = sns.light_palette("blue", as_cmap=True)
display(df[num_features].describe().T.style.background_gradient(cmap=cm))

<div style="
    background: linear-gradient(135deg, #1f2a44, #2c3e50);
    color: #ecf0f1;
    padding: 25px;
    border-radius: 12px;
    margin: 25px 0;
    box-shadow: 0 6px 18px rgba(0,0,0,0.25);
">

<div style="
    color:#f9e79f; 
    margin: 0 0 20px 0; 
    font-size: 22px; 
    font-weight: 600;
    letter-spacing: 0.5px;
">
📊 Descriptive Statistics
</div>

<ul style="line-height: 1.8; padding-left: 20px;">

<li><code>person_age</code>: Avg ~27.7; median ~26; range (~20 → 144) → mostly young population with extreme outliers (likely data quality issues).</li>

<li><code>person_income</code>: Avg ~66K; median ~55K; very wide range (~4K → 6M) → highly skewed distribution with strong outliers.</li>

<li><code>person_emp_length</code>: Avg ~4.8 years; median ~4; range (~0 → 123) → reasonable spread but contains unrealistic high values.</li>

<li><code>loan_amnt</code>: Avg ~9.6K; median ~8K; range (~500 → 35K) → moderately distributed loan sizes.</li>

<li><code>loan_int_rate</code>: Avg ~11%; range (~5.4 → 23.2) → typical lending rates with moderate variability.</li>

<li><code>loan_percent_income</code>: Avg ~17%; range (~0 → 83%) → indicates some borrowers take disproportionately high loans relative to income.</li>

<li><code>loan_to_income_ratio</code>: Similar distribution to <code>loan_percent_income</code> → potential redundancy between features.</li>

<li><code>cb_person_cred_hist_length</code>: Avg ~5.8 years; range (~2 → 30) → relatively short credit history for most borrowers.</li>

<li><code>other_debt</code>: Avg ~11.5K; extremely wide range (~225 → 1.18M) → highly skewed with significant outliers.</li>

<li><code>debt_to_income_ratio</code>: Avg ~0.35; range (~0.06 → 1.05) → high values indicate financial stress and elevated default risk.</li>

<li><code>open_accounts</code>: Avg ~8; range (~0 → 15) → moderate credit activity across customers.</li>

<li><code>credit_utilization_ratio</code>: Avg ~0.50; range (~0.05 → 0.95) → many users utilize a large portion of their credit limit.</li>

</ul>

</div>

In [ ]:
df = Utils.reduce_mem_usage(cols=num_features, df=df)
df = Utils.convert_cat(features=cat_features, df=df)

# Display information about the DataFrames
df.info()

In [ ]:
# Data Describe
display(df[cat_features].describe().T.style.background_gradient(cmap="Blues", subset=["unique", "freq"]))

<div style="
    background: linear-gradient(135deg, #1f2a44, #2c3e50);
    color: #ecf0f1;
    padding: 25px;
    border-radius: 12px;
    margin: 25px 0;
    box-shadow: 0 6px 18px rgba(0,0,0,0.25);
">

<div style="
    color:#f9e79f; 
    margin: 0 0 20px 0; 
    font-size: 22px; 
    font-weight: 600;
    letter-spacing: 0.5px;
">
📊 Categorical Features Overview
</div>

<ul style="line-height: 1.8; padding-left: 20px;">

<li><code>person_home_ownership</code>: 4 categories; most customers are <b>RENT (~50%)</b> → suggests many borrowers do not own property, potentially increasing risk exposure.</li>

<li><code>loan_intent</code>: 6 categories; <b>EDUCATION</b> is the most common → indicates strong demand for education-related loans.</li>

<li><code>loan_grade</code>: 7 levels (A–G); majority in <b>Grade A</b> → dataset is skewed toward lower-risk borrowers.</li>

<li><code>cb_person_default_on_file</code>: Binary; majority <b>No (~82%)</b> → relatively low historical default rate.</li>

<li><code>gender</code>: Balanced distribution (Male slightly higher) → unlikely to strongly bias model.</li>

<li><code>marital_status</code>: <b>Single</b> is the dominant group → may correlate with different financial behaviors.</li>

<li><code>education_level</code>: Most borrowers have <b>High School</b> education → mid-level education dominates the dataset.</li>

<li><code>employment_type</code>: Majority are <b>Full-time</b> → indicates relatively stable income sources.</li>

<li><code>country</code>: 3 categories; <b>UK</b> appears most frequently → dataset concentrated in specific regions.</li>

<li><code>state</code>: 9 categories; moderate cardinality → useful for regional analysis.</li>

<li><code>city</code>: 18 categories; relatively higher cardinality → may require encoding strategies beyond simple one-hot.</li>

<li><code>loan_term_months</code>: 4 categories; <b>36 months</b> most common → standard loan duration dominates.</li>

<li><code>past_delinquencies</code>: 7 categories; majority have <b>0 past delinquencies</b> → indicates generally good repayment behavior.</li>

</ul>

</div>

<!-- Include Google Fonts for a modern font -->
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@700&display=swap" rel="stylesheet">

# <span style="color:transparent;">Data Quality Checks</span>

<div style="
    border-radius: 15px; 
    border: 2px solid #003366; 
    padding: 10px; 
    background: linear-gradient(135deg, #3a0ca3, #7209b7 30%, #f72585 80%);
    text-align: center; 
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.5);
">
    <h1 style="
        color: #fff;
        text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.7);
        font-weight: bold;
        margin-bottom: 10px;
        font-size: 36px;
        font-family: 'Roboto', sans-serif;
        letter-spacing: 1px;
    ">
        Data Quality Checks
    </h1>
</div>


## Missing Value

In [ ]:
logger.info("Start checking NULL values")
logger.info("-" * 80)
logger.info(f"Checking missing values for {CFG.DATASET_NAME}...")
DataValidator.checkNULL(df=df, logger=logger)
logger.info("-" * 80)
logger.info("Completed checking for NULL values across datasets.")

## Checking duplicate Value

In [ ]:
logger.info("Start checking duplicate values")
logger.info("-" * 80)
logger.info(f"Data duplicate Analysis...")
DataValidator.checkDuplicate(df=df, logger=logger)
logger.info("-" * 80)
logger.info("Duplicate value inspection successfully completed across datasets.")

## Checking Outlier Value

In [ ]:
logger.info("Start checking outlier dataset")
logger.info("-" * 80)
outlier_data = DataValidator.checkOutlier(df=df, dataset_name="Bank Marketing", 
                                          list_feature=num_features, logger=logger)
logger.info("-" * 80)
logger.info("Outlier checking successfully completed across datasets.")

<div style="
    background: linear-gradient(135deg, #1f2a44, #2c3e50);
    color: #ecf0f1;
    padding: 25px;
    border-radius: 12px;
    margin: 25px 0;
    box-shadow: 0 6px 18px rgba(0,0,0,0.25);
">

<div style="
    color:#f9e79f; 
    margin: 0 0 20px 0; 
    font-size: 22px; 
    font-weight: 600;
    letter-spacing: 0.5px;
">
📊 Data Quality Insights: Missing Values, Duplicates, and Outliers
</div>

<b>Missing Values Analysis</b>

<ul>
<li>Missing values were identified in <b>2 features</b>:</li>
<ul>
<li><code>loan_int_rate</code>: 3,116 missing (~9.56%) → significant and requires careful imputation.</li>
<li><code>person_emp_length</code>: 895 missing (~2.75%) → relatively minor but still needs handling.</li>
</ul>
<li>Total missing values: <b>4,011</b> across <b>32,581 records</b>.</li>
<li>Overall, the dataset is largely complete, with missingness concentrated in specific financial attributes.</li>
</ul>

<b>Duplicate Records Analysis</b>

<ul>
<li>No duplicate rows were detected across the dataset.</li>
<li>This ensures data integrity and prevents bias from repeated observations.</li>
</ul>

<b>Outlier Analysis</b>

<ul>
<li>Outliers are present across multiple numerical features, indicating high variability in borrower profiles.</li>
<ul>
<li>Strong outliers detected in <code>person_age</code>, <code>person_income</code>, <code>loan_amnt</code>, and <code>other_debt</code>.</li>
<li>Financial ratio features such as <code>loan_percent_income</code>, <code>loan_to_income_ratio</code>, and <code>debt_to_income_ratio</code> also show notable outliers.</li>
<li><code>person_emp_length</code> and <code>cb_person_cred_hist_length</code> contain unrealistic extreme values, suggesting potential data quality issues.</li>
</ul>
<li>Some features like <code>open_accounts</code> and <code>credit_utilization_ratio</code> show stable distributions with <b>no outliers detected</b>.</li>
<li>Outliers in financial datasets often reflect real-world extreme cases (e.g., high debt or income), so <b>removal should be handled carefully</b>.</li>
</ul>

</div>

<!-- Include Google Fonts for a modern font -->
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@700&display=swap" rel="stylesheet">

# <span style="color:transparent;">Exploratory Data Analysis (EDA)</span>

<div style="
    border-radius: 15px; 
    border: 2px solid #003366; 
    padding: 10px; 
    background: linear-gradient(135deg, #3a0ca3, #7209b7 30%, #f72585 80%);
    text-align: center; 
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.5);
">
    <h1 style="
        color: #fff;
        text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.7);
        font-weight: bold;
        margin-bottom: 10px;
        font-size: 36px;
        font-family: 'Roboto', sans-serif;
        letter-spacing: 1px;
    ">
        Exploratory Data Analysis (EDA)
    </h1>
</div>

## Loan Status Distribution

In [ ]:
mapping = {0: "Non-Default", 1: "Default"}
df["loan_status"] = df["loan_status"].map(mapping)

order = ["Default", "Non-Default"]
loan_status_distribution  = df[CFG.TARGET_FEATURE].value_counts().reindex(order)
fig, ax = plt.subplots(nrows=1, ncols=2, sharey=False, figsize=(12, 5))
n_color = ColorUtils.color(n_colors=df[CFG.TARGET_FEATURE].nunique())
# ax[0]
sns.countplot(data=df, x = CFG.TARGET_FEATURE, palette=n_color, 
              ax=ax[0], order=order)
ax[0].set_title(f"Loan Status Distribution - {CFG.DATASET_NAME}", weight="bold", fontsize=12, pad=15)
for container in ax[0].containers:
    ax[0].bar_label(container, fmt="%d", label_type="edge", fontsize=10)
ax[0].set_ylabel("Number of Customer")
ax[0].set_xlabel("")
sns.despine(ax=ax[0], top=True, right=True, left=False, bottom=False)

# ax[1]
ax[1].pie(
    loan_status_distribution,
    labels=order,
    colors = n_color,
    autopct = "%1.2f%%",
    startangle = 150,
    explode = (0, 0.08),
    shadow= True
)
ax[1].set_title(f"Loan Status Breakdown - {CFG.DATASET_NAME}",  weight="bold", fontsize=12, pad=15)

plt.tight_layout()
plt.subplots_adjust(hspace=0.3, wspace=0.2)
plt.show()

<div style="
    background: linear-gradient(135deg, #1f2a44, #2c3e50);
    color: #ecf0f1;
    padding: 25px;
    border-radius: 12px;
    margin: 25px 0;
    box-shadow: 0 6px 18px rgba(0,0,0,0.25);
">

<div style="
    color:#f9e79f; 
    margin: 0 0 20px 0; 
    font-size: 22px; 
    font-weight: 600;
    letter-spacing: 0.5px;
">
📊 Insight
</div>

<table>
<tr>
<th>Dataset</th>
<th>Non-Default (0)</th>
<th>Default (1)</th>
<th>Default Rate</th>
</tr>
<tr>
<td><b>Credit Risk</b></td>
<td>25,473</td>
<td>7,108</td>
<td><b>21.82%</b></td>
</tr>
</table>

<ul style="line-height: 1.8; padding-left: 20px;">

<li><b>Majority class</b>: <code>Non-Default</code> (~78.18%) → most customers successfully repay their loans.</li>

<li><b>Minority class</b>: <code>Default</code> (~21.82%) → represents borrowers with repayment risk.</li>

<li>The dataset shows a <b>moderate class imbalance</b>, with roughly a <b>1:4 ratio</b> between default and non-default cases.</li>

<li>The default rate (~22%) is <b>significantly higher</b> than typical banking datasets → providing a stronger signal for model learning.</li>

<li>This imbalance suggests the need for:
<ul>
<li>Evaluation metrics beyond accuracy (e.g., <b>ROC-AUC, F1-score</b>)</li>
<li>Potential use of <b>class weighting</b> or <b>resampling techniques</b></li>
</ul>
</li>

</ul>

</div>

## Numerical Feature Distributions

In [ ]:
def plot_numerical_features(df=df, numerical_features = num_features):
    """
    Plot distribution of numerical features using Histogram and Boxplot.
    """

    # Create subplot grid
    fig, ax = plt.subplots(len(numerical_features), 2, figsize=(12, len(numerical_features)*4))
    # Ensure ax is always 2D (important if only 1 feature)
    ax = np.atleast_2d(ax)

    for i, feature in enumerate(numerical_features):
        # Histogram + KDE
        sns.histplot(data=df[feature], color="#4C72B0", bins = 20, kde=True, ax=ax[i, 0])
        ax[i, 0].set_title(f"Histogram of {feature.title()}", weight="bold", fontsize=12, pad=15)
        ax[i, 0].set_xlabel("")
        ax[i, 0].set_ylabel("")

        # Draw vertical line for median
        ax[i, 0].axvline(df[feature].median(), color="#C44E52", linestyle="--", label=f"Median: {df[feature].median()}")
        ax[i, 0].legend(loc="upper right")
        sns.despine(ax=ax[i, 0], top=True, right=True, left=False, bottom=False)

        # Horizontal Boxplot
        sns.boxplot(data=df[feature], color="#55A868", orient="h", ax=ax[i, 1])
        ax[i, 1].set_title(f"Horizontal Boxplot of {feature.title()}", weight="bold", fontsize=12, pad=15)
        ax[i, 1].set_xlabel("")
        sns.despine(ax=ax[i, 1], top=True, right=True, left=False, bottom=False)

    # Adjust layout to prevent overlap
    plt.tight_layout()
    plt.show()

plot_numerical_features()

In [ ]:
skew_feature = DataValidator.check_skewness(data=df, dataset_name= CFG.DATASET_NAME, 
                                            numerical_features=num_features, logger=logger)
logger.info("-" * 80)
logger.info("Checking skewness successfully completed the dataset.")

<div style="
    background: linear-gradient(135deg, #1f2a44, #2c3e50);
    color: #ecf0f1;
    padding: 25px;
    border-radius: 12px;
    margin: 25px 0;
    box-shadow: 0 6px 18px rgba(0,0,0,0.25);
">

<div style="
    color:#f9e79f; 
    margin: 0 0 20px 0; 
    font-size: 22px; 
    font-weight: 600;
    letter-spacing: 0.5px;
">
📊 Insight
</div>

<ul style="line-height: 1.8; padding-left: 20px;">

<li>Most numerical features in the dataset are <b>right-skewed</b>, indicating that values are concentrated at lower ranges with <b>long tails and extreme outliers</b>.</li>

<li><b>9 out of 12 features</b> show significant skewness (<code>|skew| > 0.5</code>), highlighting strong non-normal distributions.</li>

<li>Features such as <code>person_income</code> and <code>other_debt</code> exhibit <b>extreme skewness</b> (skew &gt; 25), suggesting the presence of very large outliers and heavy-tailed distributions.</li>

<li>Moderately skewed features include <code>debt_to_income_ratio</code>, indicating some imbalance but less extreme compared to financial amount variables.</li>

<li>Features like <code>loan_int_rate</code>, <code>open_accounts</code>, and <code>credit_utilization_ratio</code> are <b>approximately symmetric</b>, making them more stable for modeling.</li>

<li>The strong skewness in financial variables reflects real-world behavior, where a small number of customers have <b>exceptionally high income or debt levels</b>.</li>

<li><b>Data transformation techniques</b> such as <b>log transformation, Box-Cox, or Yeo-Johnson</b> are recommended to reduce skewness and improve model performance.</li>

</ul>

</div>

## Correlation Analysis of Numerical Features

In [ ]:
corr_matrix = df[num_features].corr(numeric_only=True)
# one_like can build a matrix of boolean(True, False) with the same shape as our data
ones_corr = np.ones_like(corr_matrix, dtype=bool)
mask = np.triu(ones_corr)
adjusted_mask = mask[1:, :-1]
adjusted_cereal_corr = corr_matrix.iloc[1:, :-1]

fig, ax = plt.subplots(figsize = (12, 8))
# That method uses HUSL colors, so you need hue, saturation, and lightness. 
# I used hsluv.org to select the colors of this chart.
cmap = sns.diverging_palette(0, 230, 90, 60, as_cmap=True)

sns.heatmap(data=adjusted_cereal_corr, mask=adjusted_mask,
            annot=True, fmt=".2f", cmap=cmap,
            vmin=-1, vmax=1, linecolor="white", linewidths=0.5)

title = f"Correlation Heatmap of {CFG.DATASET_NAME}"
ax.set_title(title, loc="center", fontsize=12, weight="bold", pad=15)
plt.tight_layout()
plt.show()

<div style="
    background: linear-gradient(135deg, #1f2a44, #2c3e50);
    color: #ecf0f1;
    padding: 25px;
    border-radius: 12px;
    margin: 25px 0;
    box-shadow: 0 6px 18px rgba(0,0,0,0.25);
">

<div style="
    color:#f9e79f; 
    margin: 0 0 20px 0; 
    font-size: 22px; 
    font-weight: 600;
    letter-spacing: 0.5px;
">
📊 Correlation Insights
</div>

<ul style="line-height: 1.8; padding-left: 20px;">

<li>Most numerical variables exhibit <b>weak correlations</b>, indicating that customer financial behaviors are largely independent.</li>

<li>A <b>very strong correlation</b> is observed between <code>loan_percent_income</code> and <code>loan_to_income_ratio</code> (~1.00), confirming that these features are <b>redundant</b> and carry nearly identical information.</li>

<li><code>debt_to_income_ratio</code> shows a <b>strong correlation</b> (~0.83) with both <code>loan_percent_income</code> and <code>loan_to_income_ratio</code>, indicating overlapping financial risk signals.</li>

<li><code>person_income</code> and <code>other_debt</code> are also <b>highly correlated</b> (~0.89), suggesting that higher-income individuals tend to carry higher absolute debt levels.</li>

<li>A strong relationship between <code>person_age</code> and <code>cb_person_cred_hist_length</code> (~0.86) reflects expected real-world behavior, as older individuals typically have longer credit histories.</li>

<li>Moderate correlations exist between <code>loan_amnt</code> and financial ratios (~0.48–0.58), indicating that larger loans are associated with higher financial burden.</li>

<li>Features such as <code>open_accounts</code> and <code>credit_utilization_ratio</code> show <b>near-zero correlation</b> with other variables, suggesting they provide <b>unique, independent signals</b> for modeling.</li>

<li>Overall, while most features are independent, a few clusters of correlated variables suggest potential <b>multicollinearity</b>, which should be addressed through feature selection or transformation.</li>

</ul>

</div>

## Categorical Feature Distributions

In [ ]:
def plot_categorical_distribution_across_datasets(df: pd.DataFrame, feature: str) -> None:
    """
    Display distribution summary of a categorical feature with styled table.

    Parameters
    ----------
    df : pd.DataFrame
        Input dataset.
    feature : str
        Categorical column to analyze.
    """

    # --- Summarize categorical feature ---
    summary = (
        df[feature]
        .value_counts()
        .rename_axis("Category")
        .reset_index(name="Data_Count")
    )

    # Calculate percentage distribution
    summary["Data_%"] = round(summary["Data_Count"] / summary["Data_Count"].sum() * 100, 2)

    # Ensure Category column is string
    summary["Category"] = summary["Category"].astype(str)

    # Fill NA if any
    summary[["Data_Count", "Data_%"]] = summary[["Data_Count", "Data_%"]].fillna(0)

    # ---- Styling with color gradient ----
    styled_table = (
        summary.style
        .background_gradient(cmap="Blues", subset=["Data_Count"])
        .background_gradient(cmap="Blues", subset=["Data_%"])
        .format({
            "Data_Count": "{:,}",
            "Data_%": "{:.2f}%"
        })
        .set_caption(f"Categorical Feature Distribution: {feature.title()}")
        .set_table_styles([
            {"selector": "caption", "props": [("font-size", "14px"), ("font-weight", "bold")]},
            {"selector": "th", "props": [("text-align", "center")]},
            {"selector": "td", "props": [("text-align", "center")]}
        ])
    )

    display(styled_table)
    print("*" * 80)

# ----- Run for all categorical features -----
for feature in cat_features:
    plot_categorical_distribution_across_datasets(df=df, feature=feature)

<div style="
    background: linear-gradient(135deg, #1f2a44, #2c3e50);
    color: #ecf0f1;
    padding: 25px;
    border-radius: 12px;
    margin: 25px 0;
    box-shadow: 0 6px 18px rgba(0,0,0,0.25);
">

<div style="
    color:#f9e79f; 
    margin: 0 0 20px 0; 
    font-size: 22px; 
    font-weight: 600;
    letter-spacing: 0.5px;
">
📊 Categorical Feature Distributions
</div>

<ul style="line-height: 1.8; padding-left: 20px;">

<li>The dataset is dominated by <b>non-homeowners</b>, with most customers either <b>RENT (~50%)</b> or <b>MORTGAGE (~41%)</b>, indicating moderate financial leverage.</li>

<li>Loan purposes are well distributed, with <b>education, medical, and venture</b> loans being the most common (~17–20% each), suggesting diverse borrowing needs.</li>

<li>The majority of loans fall into <b>high-quality grades (A & B ~65%)</b>, indicating that most borrowers are considered low-risk by lending standards.</li>

<li>Most customers have <b>no prior default history (~82%)</b>, reinforcing a generally reliable borrower base.</li>

<li>The dataset is <b>gender-balanced</b> (~50/50), minimizing gender bias in modeling.</li>

<li><b>Single individuals (~50%)</b> dominate the dataset, followed by married customers (~35%), suggesting varying financial responsibilities.</li>

<li>Most borrowers have <b>mid-level education</b>, with <b>High School (~40%)</b> and <b>Bachelor (~35%)</b> being the most common.</li>

<li><b>Full-time employment (~60%)</b> is the dominant employment type, indicating stable income for the majority of customers.</li>

<li>Geographically, the dataset is <b>evenly distributed across countries (UK, USA, Canada)</b>, providing balanced regional representation.</li>

<li>State and city distributions are relatively <b>uniform</b>, suggesting no strong geographic concentration or bias.</li>

<li><b>Loan terms</b> are concentrated around <b>36 months (~40%)</b> and <b>60 months (~30%)</b>, reflecting standard lending durations.</li>

<li>Most customers have <b>no past delinquencies (~60%)</b>, while a smaller portion (~30%) has only <b>1 prior delinquency</b>, indicating generally good repayment behavior.</li>

</ul>

</div>

## Bivariate Analysis

In [ ]:
for feature in num_features:
    display(HTML(f"<h2 style='text-align:center; font-size:22px; color:blue;'><b>Distribution of {feature.title()} by Loan Status</b></h2>"))
    StatUtils.analyze_numeric_feature_by_target(feature=feature, df=df, target_feature=CFG.TARGET_FEATURE, logger=logger)

<div style="
    background: linear-gradient(135deg, #1f2a44, #2c3e50);
    color: #ecf0f1;
    padding: 25px;
    border-radius: 12px;
    margin: 25px 0;
    box-shadow: 0 6px 18px rgba(0,0,0,0.25);
">

<div style="
    color:#f9e79f; 
    margin: 0 0 20px 0; 
    font-size: 22px; 
    font-weight: 600;
    letter-spacing: 0.5px;
">
📊 Feature-Level Insights
</div>

<ul style="line-height: 1.8; padding-left: 20px;">

<li><b>Person_Age</b>: Age is nearly identical between the two groups (mean ~27), so it is <b>not a strong discriminator</b> for default.  
Although statistically significant (p-value ≈ 0), the <b>effect size is very small → negligible practical impact</b>.</li>

<li><b>Person_Income</b>: Clear difference — <b>Non-default customers have significantly higher income (~70k vs ~49k)</b> → income is an <b>important feature</b>.</li>

<li><b>Person_Emp_Length</b>: Slight difference — <b>Non-default customers have longer employment (~5 years vs ~4 years)</b> → weak signal.  
Despite statistical significance, the <b>effect size is negligible</b> and the distribution is highly skewed → low practical impact.</li>

<li><b>Loan_Amnt</b>: Loan amount is <b>higher in the Default group (~10.8k vs ~9.2k)</b> → larger loans are associated with higher risk.</li>

<li><b>Loan_Int_Rate</b>: Clear difference — <b>Default group has higher interest rates (~13% vs ~10.4%)</b> → <b>very strong feature</b>.  
Large effect size (d≈0.83) → <b>key driver of default risk</b>.</li>

<li><b>Loan_Percent_Income</b>: <b>Higher in Default (~0.25 vs ~0.15)</b> → borrowers exceed repayment capacity more often → <b>important feature</b>.  
Medium effect size.</li>

<li><b>Loan_To_Income_Ratio</b>: Very similar to the previous feature → <b>important but redundant</b> with <code>loan_percent_income</code>.  
Medium effect size.</li>

<li><b>Cb_Person_Cred_Hist_Length</b>: Nearly identical between groups (~5.8 vs ~5.7) → <b>not a strong discriminator</b>.  
Despite statistical significance, <b>effect size is very small → low practical impact</b>.</li>

<li><b>Other_Debt</b>: Counterintuitive result — <b>Non-default has higher total debt (~12.3k vs ~8.6k)</b> → absolute debt is <b>not a reliable risk indicator</b>.  
Highly skewed with extreme outliers → <b>low practical value</b>; ratios are more meaningful.</li>

<li><b>Debt_To_Income_Ratio</b>: <b>Higher in Default (~0.42 vs ~0.32)</b> → financial burden is a key risk factor.  
Large effect size (d≈0.75) → <b>one of the most important features</b>.</li>

<li><b>Open_Accounts</b>: Nearly identical (~8) → <b>no meaningful difference</b>.  
p-value > 0.05 and effect size ≈ 0 → <b>not predictive</b>.</li>

<li><b>Credit_Utilization_Ratio</b>: Nearly identical (~0.50) → <b>no meaningful difference</b>.  
p-value > 0.05 and effect size ≈ 0 → <b>does not contribute to prediction</b>.</li>

</ul>

</div>

In [ ]:
for feature in cat_features:
    StatUtils.analyze_categorical_feature_vs_target(cat = feature, target_feature=CFG.TARGET_FEATURE,
                                                    logger=logger, df=df)

<div style="
    background: linear-gradient(135deg, #1f2a44, #2c3e50);
    color: #ecf0f1;
    padding: 25px;
    border-radius: 12px;
    margin: 25px 0;
    box-shadow: 0 6px 18px rgba(0,0,0,0.25);
">

<div style="
    color:#f9e79f; 
    margin: 0 0 20px 0; 
    font-size: 22px; 
    font-weight: 600;
    letter-spacing: 0.5px;
">
📊 Categorical Feature Insights
</div>

<ul style="line-height: 1.8; padding-left: 20px;">

<li><b>person_home_ownership</b>: Renters are the <b>highest-risk group (~32% default)</b>, while homeowners (especially <b>OWN ~7%</b>) are the safest → <b>strong risk indicator</b>.</li>

<li><b>Loan_Intent</b>: Clearly associated with default → <b>DEBTCONSOLIDATION &amp; MEDICAL</b> have higher risk, while <b>EDUCATION &amp; VENTURE</b> are safer segments.</li>

<li><b>Loan_Grade</b>: <b>Strongest risk driver</b> → default rate increases sharply from <b>A (~10%) → G (~98%)</b>.</li>

<li><b>Cb_Person_Default_On_File</b>: Prior default (<b>Y</b>) has much higher risk (<b>~38% vs ~18%</b>) → <b>strong and intuitive predictor</b>.</li>

<li><b>Gender</b>: Default rates are identical (~22%), p-value &gt; 0.05 → <b>no predictive power</b>.</li>

<li><b>Marital_Status</b>: Nearly identical (~21–22%), p-value &gt; 0.05 → <b>no meaningful relationship</b>.</li>

<li><b>Education_Level</b>: Very similar (~21–22%), p-value &gt; 0.05 → <b>no meaningful impact</b>.</li>

<li><b>Employment_Type</b>: Nearly identical (~22–23%), p-value &gt; 0.05 → <b>no meaningful impact</b>.</li>

<li><b>Country</b>: Almost identical (~22%), p-value &gt; 0.05 → <b>no meaningful impact</b>.</li>

<li><b>State</b>: Very similar (~21–23%), p-value &gt; 0.05 → <b>no meaningful impact</b>.</li>

<li><b>City</b>: Consistent (~20–24%), p-value &gt; 0.05 → <b>no meaningful impact</b>.</li>

<li><b>Loan_Term_Months</b>: Statistically significant but weak (p &lt; 0.05), with similar default rates (~20–22%).</li>

<li><b>Past_Delinquencies</b>: No significant relationship (p-value &gt; 0.05), default rates remain consistent (~21–25%).</li>

</ul>

</div>

<!-- Include Google Fonts for a modern font -->
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@700&display=swap" rel="stylesheet">

# <span style="color:transparent;">Overall Picture</span>

<div style="
    border-radius: 15px; 
    border: 2px solid #003366; 
    padding: 10px; 
    background: linear-gradient(135deg, #3a0ca3, #7209b7 30%, #f72585 80%);
    text-align: center; 
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.5);
">
    <h1 style="
        color: #FFFFFF; 
        text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.7); 
        font-weight: bold; 
        margin-bottom: 5px; 
        font-size: 28px; 
        font-family: 'Roboto', sans-serif;
        letter-spacing: 1px;
    ">
        Overall Picture
    </h1>
</div>

<div style="
    background: linear-gradient(135deg, #1f2a44, #2c3e50);
    color: #ecf0f1;
    padding: 25px;
    border-radius: 12px;
    margin: 25px 0;
    box-shadow: 0 6px 18px rgba(0,0,0,0.25);
">

<div style="
    color:#f9e79f; 
    margin: 0 0 20px 0; 
    font-size: 22px; 
    font-weight: 600;
    letter-spacing: 0.5px;
">
📊 Overall Picuture
</div>

| Feature                                                                | Key Insight                                                                                                                  |
| ---------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------- |
| **loan_grade**                                                         | **Clear risk segmentation** → lower grades (D–G) have much higher default rates → useful for approval and pricing decisions. |
| **loan_int_rate**                                                      | **Higher interest rates indicate higher risk borrowers** → strong signal of default likelihood.                              |
| **debt_to_income_ratio**                                               | **Core risk driver** → customers with higher financial burden are much more likely to default.                               |
| **loan_percent_income**                                                | **Repayment pressure matters** → higher loan-to-income ratio increases default risk.                                         |
| **person_income**                                                      | **Higher income = safer borrowers** → strong indicator of repayment ability.                                                 |
| **person_home_ownership**                                              | **Housing stability matters** → renters are riskier, while homeowners are significantly safer.                               |
| **cb_person_default_on_file**                                          | **Past behavior predicts future risk** → customers with prior default are much more likely to default again.                 |
| **loan_intent**                                                        | **Loan purpose impacts risk** → debt consolidation & medical loans are riskier; education & venture are safer.               |
| **loan_amnt**                                                          | **Larger loans increase exposure** → slightly higher default risk.                                                           |
| **loan_to_income_ratio**                                               | **Overlaps with loan_percent_income** → similar signal, should keep one to avoid redundancy.                                 |
| **loan_term_months**                                                   | **Minimal impact** → default rates are similar across different loan durations.                                              |
| **person_age**                                                         | **Low business impact** → similar behavior across age groups.                                                                |
| **person_emp_length**                                                  | **Weak signal of stability** → longer employment slightly reduces risk but not significant.                                  |
| **cb_person_cred_hist_length**                                         | **Limited usefulness** → credit history length alone does not explain default well.                                          |
| **other_debt**                                                         | **Misleading signal** → absolute debt is less meaningful than ratios due to skew.                                            |
| **open_accounts**                                                      | **No meaningful impact** → similar behavior across groups.                                                                   |
| **credit_utilization_ratio**                                           | **No clear signal** → does not differentiate default behavior.                                                               |
| **past_delinquencies**                                                 | **Weak predictor** → less informative than prior default flag.                                                               |
| **gender / marital_status / education_level / country / state / city** | **No business impact** → default rates are consistent across demographic and geographic groups.                              |

</div>


<!-- Include Google Fonts for a modern font -->
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@700&display=swap" rel="stylesheet">

# <span style="color:transparent;">Data Pre-processing</span>

<div style="
    border-radius: 15px; 
    border: 2px solid #003366; 
    padding: 10px; 
    background: linear-gradient(135deg, #3a0ca3, #7209b7 30%, #f72585 80%);
    text-align: center; 
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.5);
">
    <h1 style="
        color: #FFFFFF; 
        text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.7); 
        font-weight: bold; 
        margin-bottom: 5px; 
        font-size: 28px; 
        font-family: 'Roboto', sans-serif;
        letter-spacing: 1px;
    ">
        Data Pre-processing
    </h1>
</div>

## Handle skewness

In [ ]:
logger.info("-" * 80)
skew_feature = DataValidator.check_skewness(data=df, dataset_name=CFG.DATASET_NAME, 
                                            numerical_features=num_features, logger=logger)
logger.info("-" * 80)
logger.info("Checking skewness successfully completed the dataset.")

In [ ]:
processed_df, transformed_columns, sparse_columns, skewed_columns = FeatureUtils.fea_HandleSkewedFeatures(df=df, num_features=skew_feature, logger=logger)

In [ ]:
num_features = ["PT_person_age", "PT_person_income", "PT_person_emp_length", "PT_loan_amnt", "loan_int_rate", "PT_loan_percent_income", 
                "PT_loan_to_income_ratio", "PT_cb_person_cred_hist_length", "PT_other_debt",  "PT_debt_to_income_ratio", "open_accounts", 
                "credit_utilization_ratio"]
logger.info("-" * 80)
skew_feature = DataValidator.check_skewness(data=processed_df, dataset_name=CFG.DATASET_NAME, 
                                            numerical_features=num_features, logger=logger)
logger.info("-" * 80)
logger.info("Checking skewness successfully completed the dataset.")

<div style="
    background: linear-gradient(135deg, #1f2a44, #2c3e50);
    color: #ecf0f1;
    padding: 25px;
    border-radius: 12px;
    margin: 25px 0;
    box-shadow: 0 6px 18px rgba(0,0,0,0.25);
">

<div style="
    color:#f9e79f; 
    margin: 0 0 20px 0; 
    font-size: 22px; 
    font-weight: 600;
    letter-spacing: 0.5px;
">
📊 Insight
</div>

<b>Observations from the Skewness Analysis:</b>

<b>Before Transformation:</b>

<ul style="line-height: 1.8; padding-left: 20px;">
<li>Most numerical features were <b>heavily right-skewed</b>, especially <code>person_income</code> and <code>other_debt</code> (extreme skew &gt; 25).</li>
<li>A total of <b>9 out of 12 features</b> showed significant skewness (|skew| &gt; 0.5), indicating <b>non-normal distributions and presence of outliers</b>.</li>
<li>Only a few features such as <code>loan_int_rate</code>, <code>open_accounts</code>, and <code>credit_utilization_ratio</code> were approximately symmetric.</li>
</ul>

<b>After Transformation (Yeo-Johnson / Power Transform):</b>

<ul style="line-height: 1.8; padding-left: 20px;">
<li>All transformed features (<code>PT_*</code>) became <b>approximately symmetric</b> (skew ≈ 0).</li>
<li><b>No highly skewed features remain</b>, indicating a successful normalization process.</li>
<li>Extreme skewness in key financial variables (<code>person_income</code>, <code>other_debt</code>) was <b>effectively corrected</b>.</li>
<li>Original symmetric features (<code>loan_int_rate</code>, <code>open_accounts</code>, <code>credit_utilization_ratio</code>) remained stable.</li>
</ul>

<b>Conclusion:</b>

<ul style="line-height: 1.8; padding-left: 20px;">
<li>The transformation significantly <b>improved feature distributions</b>, reducing the impact of outliers.</li>
<li>This enhances <b>model stability, convergence, and performance</b>, especially for linear and distance-based models.</li>
<li>The dataset is now better suited for <b>machine learning modeling</b> with reduced bias from skewed features.</li>
</ul>

</div>

## Re-check Outlier

In [ ]:
logger.info("Start checking outlier dataset")
logger.info("-" * 80)
outlier_data = DataValidator.checkOutlier(df=processed_df, dataset_name=CFG.DATASET_NAME, 
                                          list_feature=num_features, logger=logger)
logger.info("-" * 80)
logger.info("Outlier checking successfully completed across datasets.")

## Create Test Set

In [ ]:
X = processed_df.drop(columns=CFG.TARGET_FEATURE)
y = processed_df[CFG.TARGET_FEATURE].map({"Non-Default": 0, "Default": 1})

split = StratifiedShuffleSplit(n_splits=CFG.SPLIT["n_splits"], test_size=CFG.SPLIT["test_size"], 
                               random_state=CFG.RANDOM_STATE)

for train_idx, test_idx in split.split(X, y):
    X_train = X.iloc[train_idx]
    X_test = X.iloc[test_idx]
    y_train = y.iloc[train_idx]
    y_test = y.iloc[test_idx]

## Scale and Encode

In [ ]:
list_feature_num_stand = ["PT_person_age", "PT_loan_percent_income", "PT_loan_to_income_ratio", "PT_cb_person_cred_hist_length", "open_accounts", 
                          "credit_utilization_ratio"]
list_feature_num_robust = ["PT_person_income", "PT_person_emp_length", "PT_loan_amnt", "loan_int_rate", "PT_other_debt", 
                           "PT_debt_to_income_ratio"]

num_robust_transformer = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", RobustScaler())
])

num_stand_transformer = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])

cat_onehot_transformer = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("encoder", OneHotEncoder(handle_unknown="ignore"))
])

preprocessor = ColumnTransformer(transformers=[
    ("num_robust", num_robust_transformer, list_feature_num_robust),
    ("num_standard", num_stand_transformer, list_feature_num_stand),
    ("cat_onehot", cat_onehot_transformer, cat_features)
])

preprocessor.fit(X_train)
X_train_prepared = preprocessor.transform(X_train)
list_feature_prepared = preprocessor.get_feature_names_out().tolist()
clean_features = [col.replace("num_standard__", "").replace("num_robust__", "").replace("cat_onehot__", "").replace("PT_", "") for col in list_feature_prepared]

<!-- Include Google Fonts for a modern font -->
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@700&display=swap" rel="stylesheet">

# <span style="color:transparent;">Metric</span>

<div style="
    border-radius: 15px; 
    border: 2px solid #003366; 
    padding: 10px; 
    background: linear-gradient(135deg, #3a0ca3, #7209b7 30%, #f72585 80%);
    text-align: center; 
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.5);
">
    <h1 style="
        color: #FFFFFF; 
        text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.7); 
        font-weight: bold; 
        margin-bottom: 5px; 
        font-size: 28px; 
        font-family: 'Roboto', sans-serif;
        letter-spacing: 1px;
    ">
        Metric
    </h1>
</div>


**AUC-ROC Overview**

**AUC-ROC** evaluates a model’s ability to **distinguish between two classes**.

* **ROC curve** plots **TPR vs FPR**
* **AUC** = Area under ROC curve → **ranking quality of predictions**

**Why AUC-ROC?**

* Robust to **class imbalance**
* Measures **ranking ability**, not fixed threshold (0.5)
* Common metric in **credit risk & fraud prediction**

> Higher AUC → better separation between *Paid* vs *Default*

**Interpretation**

| AUC Score     | Meaning                  |
| ------------- | ------------------------ |
| **1.0**       | Perfect model            |
| **0.9+**      | Excellent discrimination |
| **0.7 – 0.9** | Good                     |
| **0.5**       | Random guessing          |
| **< 0.5**     | Worse than random        |

(Source: [ROC AUC](https://en.wikipedia.org/wiki/Receiver_operating_characteristic))

<!-- Include Google Fonts for a modern font -->
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@700&display=swap" rel="stylesheet">

# <span style="color:transparent;">Modeling</span>

<div style="
    border-radius: 15px; 
    border: 2px solid #003366; 
    padding: 10px; 
    background: linear-gradient(135deg, #3a0ca3, #7209b7 30%, #f72585 80%);
    text-align: center; 
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.5);
">
    <h1 style="
        color: #FFFFFF; 
        text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.7); 
        font-weight: bold; 
        margin-bottom: 5px; 
        font-size: 28px; 
        font-family: 'Roboto', sans-serif;
        letter-spacing: 1px;
    ">
        Modeling
    </h1>
</div>

## CatBoostClassifier

In [ ]:
study_cat = optuna.create_study(direction="maximize")
study_cat.optimize(
    lambda trial: objective_catboost(trial, X_train_prepared, y_train, logger),
    n_trials=10, show_progress_bar=True
)

In [ ]:
model_cb = CatBoostClassifier(**study_cat.best_params, loss_function="Logloss", auto_class_weights="Balanced", random_seed=CFG.RANDOM_STATE,
                              verbose = 0, thread_count=-1, eval_metric="AUC", grow_policy="SymmetricTree", task_type="CPU",
                              allow_writing_files = False)
X_test_prepared = preprocessor.transform(X_test)
oof_pred_cb, test_pred_cb, fold_auc_cb, fold_models_cb = run_oof(model = model_cb, X = X_train_prepared, y = y_train.values, 
                                                                 X_test = X_test_prepared, y_test=y_test, logger=logger)

## LGBMClassifier

In [ ]:
study_lgbm = optuna.create_study(direction="maximize")
study_lgbm.optimize(lambda trial: objective_lgbm(trial, X_train_prepared, y_train, logger),
    n_trials=10, show_progress_bar=True
)

In [ ]:
model_lgbm = LGBMClassifier(**study_lgbm.best_params, objective="binary", metric="auc",
                            verbosity=-1, n_jobs=-1, boosting_type="gbdt",
                            class_weight="balanced", random_state=CFG.RANDOM_STATE)
oof_pred_lgbm, test_pred_lgbm, fold_auc_lgbm, fold_models_lgbm = run_oof(model = model_lgbm, X = X_train_prepared, y = y_train.values, 
                                                                         X_test = X_test_prepared, y_test=y_test, logger=logger)

## HistGradientBoostingClassifier

In [ ]:
X_train_dense = X_train_prepared.toarray()

study_hgbc = optuna.create_study(direction="maximize")
study_hgbc.optimize(lambda trial: objective_histgb(trial, X_train_dense, y_train, logger),
    n_trials=10, show_progress_bar=True
)

In [ ]:
model_hgbc = HistGradientBoostingClassifier(**study_hgbc.best_params, n_iter_no_change=50, validation_fraction=0.1, early_stopping=True,
                                            class_weight="balanced", random_state=CFG.RANDOM_STATE)

X_test_dense = X_test_prepared.toarray()
oof_pred_hgbc, test_pred_hgbc, fold_auc_hgbc, fold_models_hgbc = run_oof(model = model_hgbc, X = X_train_dense, y = y_train.values, 
                                                                         X_test = X_test_dense, y_test=y_test, logger=logger)

<!-- Include Google Fonts for a modern font -->
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@700&display=swap" rel="stylesheet">

# <span style="color:transparent;">Evaluation</span>

<div style="
    border-radius: 15px; 
    border: 2px solid #003366; 
    padding: 10px; 
    background: linear-gradient(135deg, #3a0ca3, #7209b7 30%, #f72585 80%);
    text-align: center; 
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.5);
">
    <h1 style="
        color: #FFFFFF; 
        text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.7); 
        font-weight: bold; 
        margin-bottom: 5px; 
        font-size: 28px; 
        font-family: 'Roboto', sans-serif;
        letter-spacing: 1px;
    ">
        Evaluation
    </h1>
</div>

In [ ]:
max_index_cat = fold_auc_cb.index(max(fold_auc_cb))
max_index_lgbm = fold_auc_lgbm.index(max(fold_auc_lgbm))
max_index_hgbc = fold_auc_hgbc.index(max(fold_auc_hgbc))

list_model = [fold_models_cb[max_index_cat], fold_models_lgbm[max_index_lgbm], fold_models_hgbc[max_index_hgbc]]
shap_compare_models(models=list_model, X_test=X_test_prepared, logger=logger, feature_names=clean_features)

In [ ]:
fig, ax = plt.subplots(1, 3, figsize=(18, 5))
COLORS = ["#667eea", "#764ba2", "#a855f7", "#f093fb", "#fbc2eb"]

ax[0].hist(test_pred_cb, bins=50, color=COLORS[0], alpha=0.7, edgecolor="white")
ax[0].set_title("HistGradientBoosting OOF Predictions", fontsize=12, fontweight="bold", pad=15)
ax[0].set_xlabel("Predicted Probability", fontsize=10)
ax[0].grid(axis="y", alpha=0.3)

ax[1].hist(test_pred_lgbm, bins=50, color=COLORS[1], alpha=0.7, edgecolor="white")
ax[1].set_title("LightGBM OOF Predictions", fontsize=12, fontweight="bold", pad=15)
ax[1].set_xlabel("Predicted Probability", fontsize=10)
ax[1].grid(axis="y", alpha=0.3)

ax[2].hist(test_pred_hgbc, bins=50, color=COLORS[2], alpha=0.7, edgecolor="white")
ax[2].set_title("Catboost OOF Predictions", fontsize=12, fontweight="bold", pad=15)
ax[2].set_xlabel("Predicted Probability", fontsize=10)
ax[2].grid(axis="y", alpha=0.3)

plt.tight_layout()
plt.show()

In [ ]:
hgbc_binary = (test_pred_hgbc >= 0.5).astype(int)
lgb_binary = (test_pred_lgbm >= 0.5).astype(int)
cat_binary = (test_pred_cb >= 0.5).astype(int)

logger.info("HistGradientBoosting Classification Report:")
logger.info("\n" + classification_report(y_test, hgbc_binary, target_names=['No', 'Yes']))
logger.info("*" * 80)

logger.info("LightGBM Classification Report:")
logger.info("\n" + classification_report(y_test, lgb_binary, target_names=['No', 'Yes']))
logger.info("*" * 80)

logger.info("CatBoost Classification Report:")
logger.info("\n" + classification_report(y_test, cat_binary, target_names=['No', 'Yes']))
logger.info("*" * 80)

<!-- Include Google Fonts for a modern font -->
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@700&display=swap" rel="stylesheet">

# <span style="color:transparent;">Recommendation</span>

<div style="
    border-radius: 15px; 
    border: 2px solid #003366; 
    padding: 10px; 
    background: linear-gradient(135deg, #3a0ca3, #7209b7 30%, #f72585 80%);
    text-align: center; 
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.5);
">
    <h1 style="
        color: #FFFFFF; 
        text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.7); 
        font-weight: bold; 
        margin-bottom: 5px; 
        font-size: 28px; 
        font-family: 'Roboto', sans-serif;
        letter-spacing: 1px;
    ">
        Recommendation
    </h1>
</div>

<div style="
    background: linear-gradient(135deg, #1f2a44, #2c3e50);
    color: #ecf0f1;
    padding: 25px;
    border-radius: 12px;
    margin: 25px 0;
    box-shadow: 0 6px 18px rgba(0,0,0,0.25);
">

<div style="
    color:#f9e79f; 
    margin: 0 0 10px 0; 
    font-size: 22px; 
    font-weight: 600;
    letter-spacing: 0.5px;
">
📊 Key Business Recommendations
</div>

<div style="
    margin-bottom: 20px;
    font-size: 14px;
    color: #d5dbdb;
    font-style: italic;
">
Key drivers of default are <b>financial capacity</b> and <b>credit behavior</b>.
</div>

<ul style="line-height: 1.8; padding-left: 20px;">

<li>
<b>Focus on Repayment Capacity (Most Important)</b>
<ul>
<li>Use <b>DTI &amp; income ratios</b> as core approval criteria</li>
<li>High ratio → <b>reject or increase interest rate</b></li>
</ul>
</li>

<li>
<b>Leverage Loan Grade for Risk Segmentation</b>
<ul>
<li>A–B: approve &amp; low rate</li>
<li>C–D: controlled lending</li>
<li>E–G: <b>restrict / tighten conditions</b></li>
</ul>
</li>

<li>
<b>Flag High-Risk Customers</b>
<ul>
<li>Prior default (<code>cb_person_default_on_file = Y</code>) → <b>strong reject or strict policy</b></li>
<li>Renters + low income → <b>higher monitoring</b></li>
</ul>
</li>

<li>
<b>Adjust by Loan Purpose</b>
<ul>
<li>Risky: <b>debt consolidation, medical</b> → stricter approval</li>
<li>Safer: <b>education, venture</b> → growth opportunity</li>
</ul>
</li>

<li>
<b>Simplify Decision Factors</b>
<ul>
<li>Focus on: <b>income, DTI, loan grade, past default</b></li>
<li>Ignore: <b>demographics &amp; location (no impact)</b></li>
</ul>
</li>

</ul>

</div>

<!-- Include Google Fonts for a modern font -->
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@700&display=swap" rel="stylesheet">

<div style="
    border-radius: 15px; 
    border: 2px solid #003366; 
    padding: 10px; 
    background: linear-gradient(135deg, #3a0ca3, #7209b7 30%, #f72585 80%);
    text-align: center; 
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.5);
">
    <h1 style="
        color: #FFFFFF; 
        text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.7); 
        font-weight: bold; 
        margin-bottom: 5px; 
        font-size: 28px; 
        font-family: 'Roboto', sans-serif;
        letter-spacing: 1px;
    ">
        🙏 Thanks for Reading! 🚀
    </h1>
    <p style="color: #ffffff; font-size: 18px; text-align: center;">
        Happy coding 🙌😊
    </p>
</div>
